src/libmongoc/src/mongoc/mongoc-uri.c

```diff
@@ -1892,168 +1815,211 @@ mongoc_uri_finalize_directconnection (mongoc_uri_t *uri, bson_error_t *error)
+   // do anything different.
    {
-      const char *tmp;
-
-      // Only ':' is permitted among RFC-3986 gen-delims (":/?#[]@") in userinfo.
-      // However, continue supporting these characters for backward compatibility, as permitted by the Connection
-      // String spec: for backwards-compatibility reasons, drivers MAY allow reserved characters other than "@" and
-      // ":" to be present in user information without percent-encoding.
-      char *userinfo = scan_to_unichar (cursor, '@', "", &tmp);
-
-      if (userinfo) {
-         cursor = tmp + 1; // Consume userinfo delimiter.
-         bson_free (userinfo);
+      size_t userinfo_end_pos = mstr_find (remain, AT);
+      if (userinfo_end_pos == SIZE_MAX) {
+         // There is no userinfo, so we don't need to do anything speical
+         userinfo_end_pos = 0;
+      }
+      // Find the position of the first character that terminates the authority element
+      const size_t term_pos = mstr_find_first_of (remain, mstr_cstring ("/?#"), userinfo_end_pos);
+      mstr_split_at (remain, term_pos, &parts->authority, &remain);
+
+      // Now we should split the authority between the userinfo and the hosts
+      {
+         const size_t at_pos = mstr_find (parts->authority, AT);
+         if (at_pos != SIZE_MAX) {
+            // We have a userinfo component
+            mstr_split_at (parts->authority, at_pos, 1, &parts->userinfo, &parts->hosts);
+         } else {
+            // We have no userinfo, so the authority string is just the host list
+            parts->hosts = parts->authority;
+         }
       }
    }
-   // cursor -> "host1:27017,host2:27018/database?key1=value1&key2=value2"
-
-   // Find either the optional auth database delimiter or the query delimiter.
-   // e.g. "host1:27017,host2:27018/database?key1=value1&key2=value2"
-   //                              ^
-   // e.g. "host1:27017,host2:27018?key1=value1&key2=value2"
-   //                              ^
-   {
-      const char *tmp;
-
-      // Only ':', '[', and ']' are permitted among RFC-3986 gen-delims (":/?#[]@") in hostinfo.
-      const char *const terminators = "/?#@";
-
-      char *hostinfo;
 
-      // Optional auth delimiter is present.
-      if ((hostinfo = scan_to_unichar (cursor, '/', terminators, &tmp))) {
-         cursor = tmp; // Include the delimiter.
-         bson_free (hostinfo);
-      }
+   // * remain = "/path?query#fragment" (Each following component is optional, but this is the proper order)
+   const size_t path_end_pos = mstr_find_first_of (remain, mstr_cstring ("?#"));
+   mstr_split_at (remain, path_end_pos, &parts->path, &remain);
+   // * remain = "query#fragment" (Each following component is optional, but this is the proper order)
+   const size_t hash_pos = mstr_find_first_of (remain, mstr_cstring ("#"));
+   mstr_split_at (remain, hash_pos, &parts->query, &remain);
+   // * remain = "#fragment"
```

Comment:
Suggest removing special handling for `#fragment`. The connection string spec does not appear to mention a fragment. And this may trigger a backwards breaking behavior change. Example:

```c
// Previously OK. Now a parse error:
mongoc_uri_t *uri = mongoc_uri_new ("mongodb://host/?appName=bar#baz");
const char *got = mongoc_uri_get_option_as_utf8 (uri, "appName", "");
ASSERT_CMPSTR (got, "bar#baz");
mongoc_uri_destroy (uri);
```